feat: RFD 9: Spec for deployment and automation scripts

viccuad opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

The following spec is an interface specification for sharing deployment and
automation scripts of Helm charts and utility applications.


## Motivation and Context

Possible implementation for https://github.com/cloudfoundry-incubator/kubecf/issues/1012

May aid in:
https://github.com/cloudfoundry-incubator/kubecf/issues/220
https://github.com/cloudfoundry-incubator/kubecf/issues/75
https://github.com/cloudfoundry-incubator/kubecf/issues/1231
https://github.com/cloudfoundry-incubator/kubecf/issues/1232

Provides a framework (but leaves out for now):
https://github.com/cloudfoundry-incubator/kubecf/issues/1018

Relates to:
https://github.com/cloudfoundry-incubator/kubecf/issues/1422

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- An important detail to include is the version of the used cf-operator -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has security implications.
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

viccuad commented:
For future proofing, I think it would be useful to add a file `spec_ver`, containing a string with the version that the scripts implement (e.g: `0.1`). This allows us to make changes to the spec and have several charts be compliant with different spec versions in the future.

manno commented:
I'm missing context on this. Can you please add some explanation what this is for (like a user story)? 

I'm asking myself for example:

* which helm charts should implement this (kubecf?)
* who is the intended user of these scripts (ci, devs, users)? 
* where should these ten folders go? (root dir of git repo, hack, bin, ci?)
* how would someone use these folders?
* I'm also missing explanations for each of the folders. (clean what, wait for what?)

> The following spec is an interface specification for sharing deployment and automation scripts of Helm charts and utility applications.

This approach looks a bit too special to me. I'm concerned nobody else is doing it like this. 
It would replace "simply" calling helm, with executing a hierarchy of scripts?
Did you look into alternatives, maybe like [helmfile](https://github.com/roboll/helmfile)?

> Dealing with a handful of loosely dependent Helm charts and building a Helm
distribution from them without creating an all-or-nothing "super-chart" is
difficult. 

This makes me think it's about the different helm charts included in kubecf.
So how would kubecf install instructions look like with this "structure of folders and executables"?
Is this intended for releases or builds from git?
Do these scripts modify configuration (configure folder?), how is configuration distributed?
Since upstream projects won't implement this structure, it still has to be maintained, I think?

> One needs to make sure that deployment and automation scripts are
interoperable between charts, repos, repos' histories, and that they can be
reused and extended downstream as needed.

Does that mean the scripts should be used and modified by KubeCF and CAP?

> At the same time one needs to create, reuse, and extend test environments across
charts, and between upstream and downstream. The following spec is an interface specification for sharing deployment and
automation scripts of Helm charts and utility applications.

Does this also solve the problem of upstream and downstream using different clusters and testing different things? Would there be a "structure of folders and executables" for each of these use cases?

viovanov commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,176 @@
+the chart repository, but we recommend that this files are bundled inside of the
+chart: this simplifies consumption downstream.
+
+While targets are allowed to take arguments and options on execution besides the
```
Can you please define `target` before this paragraph?

viccuad replied:
done! thanks

viovanov commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,176 @@
+  already decompressed in `$pwd/$our-chart/chart/`. `$our-chart` being the name
+  of the chart at hand (e.g: kubecf).
+
+- `configure`: Outcome is a valid `$pwd/$our-chart/values.yaml` file for the
```
Why mangle yaml files? The `helm` CLI supports specifying multiple values yaml files.

viccuad replied:
That's correct, and a better approach. I think a good approach would be for `configure` target to provide a `$pwd/values/` folder with those, that the `install` and `upgrade` targets would consume then.

viccuad replied:
I'm still thinking on how to tackle this properly. Merging subsets of yaml files has the benefit that one can order then alphanumerically, and provide a deterministic way to do the merge.
I wonder about yaml files whose objects overlap. Do any charts have any of those? What happens when merging them? Does it matter?

viccuad replied:
Moved to a folder for all config values in https://github.com/cloudfoundry-incubator/kubecf/pull/1495/commits/3e5206b7c0cadbb2d136701b481830e789b34fe8

viovanov commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,176 @@
+  `$pwd/$our-chart/chart/` and upgrades the Helm Chart at hand. The upgrade must
+  work for both same version and different versions.
+
+- `clean`: Removes all possible objects created by any target from the cluster.
```
Can you split this into `clean` and `delete`?
`clean` can remove objects from the cluster, and `delete` can delete the chart.
I usually want to remove things from the cluster quite often, but rarely want to delete the chart from my local disk.

viccuad replied:
just when you were reviewing, I tackled Mario's review and reworded `clean` and added a `clobber` target that leaves the cluster pristine. Would that be ok?

viccuad replied:
renamed `clean` to `uninstall` in https://github.com/cloudfoundry-incubator/kubecf/pull/1495/commits/76c684f28d9a79c6483c15fe125b7380658fd6e9, and we have clobber nevertheless.

viovanov commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,176 @@
+- `one-offs/*`: Helper targets. Follow the same rules as the rest of targets,
+  except they may not be idempotent. E.g: `klog`, `login`, `smokes`, `cats`.
+  
+Each of these targets:
```
This section should be higher in the doc.

viccuad replied:
do you mean the one-offs or the "each of these targets" paragraph?

viccuad replied:
done in https://github.com/cloudfoundry-incubator/kubecf/pull/1495/commits/6b63f4ebbc6dbe7b849434f583953ef7d7c958cb

viovanov commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,176 @@
+  E.g: `$pwd/artifacts/cats-run.log`
+
+
+Apart from the targets, the following folders may be provided:
```
I think these _must_ be provided. They must be used to compose the helm install/upgrade commands.

viccuad replied:
Correct! fixed. Also, I added imagelist.txt.

viovanov reviewed (changes requested):
I don't think my review is finished, but would like to start discussing.
@viccuad the main concern I have is about the values yaml files, so please consider that first. 
And thanks! this is great

viccuad commented:
Thanks Mario for the review!

> I'm missing context on this. Can you please add some explanation what this is for (like a user story)?

Done! please tell me if they aren't clear.

> 
> I'm asking myself for example:
> 
>     * which helm charts should implement this (kubecf?)

Any that want to. I would start with kubecf.
Ideally, any that need to be tested together: kubecf, quarks, minibroker, stratos, stratos-metrics, caps-registry, ceph. The spec is too specific and too loose at the same time, but it's a starting point. It could completely change in follow-up versions. And at some point become irrelevant because Helm fixe, implements, or specifies the needed parts in the future.

>     * who is the intended user of these scripts (ci, devs, users)?

ci & devs. I think we are doing a disservice to users by not providing them with a Helmfile or example-values, though.

> 
>     * where should these ten folders go? (root dir of git repo, hack, bin, ci?)

It is not specified, one is free to put them wherever.

> 
>     * how would someone use these folders?

There would be a test-runner that would execute the targets. That test-runner would be simpler that current catapult.
Also, the folders can be consumed downstream by injecting the needed values.yaml subsets, or by injecting an odd `upgrade/99_apply-workaround.sh` when needed, that would get executed in the specified alphanumeric order.

> 
>     * I'm also missing explanations for each of the folders. (clean what, wait for what?)

Reworded.

> This approach looks a bit too special to me. I'm concerned nobody else is doing it like this.
> It would replace "simply" calling helm, with executing a hierarchy of scripts?
> Did you look into alternatives, maybe like [helmfile](https://github.com/roboll/helmfile)?

I have looked, but I haven't for example written a Helmfile myself for kubecf.

This RFD overlaps with Helmfiles in that it provides a declarative list of possible values for the chart, to pick and mix, which incidentally, form our test cases.
But the RFD intent is to also share implementations of CI/automation, shipping them as part of the release, to be consumed downstream. We could move to Helmfiles for the declarative values, and keep the rest of conventions for sharing the automation bits.

> This makes me think it's about the different helm charts included in kubecf.
> So how would kubecf install instructions look like with this "structure of folders and executables"?

For users, nothing changes.
For devs/CI/automation, we just change `kubecf/Makefile` to call the correct target, and this targets could be implemented with changes to the `kubecf/scripts/` folder. 

> Is this intended for releases or builds from git?

It is intended for both. Releases (just some files inside of the chart that nobody cares about) and daily git builds, which are now being consumed downstream for nightlies.

> Do these scripts modify configuration (configure folder?), how is configuration distributed?

Added explanation of configure target. Configuration is distributed and created by merging the `example-values/` files with `configure/patches`. I think we should be clearer here, or use Helmfiles.

> Since upstream projects won't implement this structure, it still has to be maintained, I think?

That could be the case. That's the status quo for Catapult, which implements more or less this for kubecf, stratos, stratos-metrics, cf-operator, and all their one-offs. Catapult would benefit from a rewrite and no implementing this spec.


> Does that mean the scripts should be used and modified by KubeCF and CAP?

Correct.

> Does this also solve the problem of upstream and downstream using different clusters and testing different things? Would there be a "structure of folders and executables" for each of these use cases?

Correct. The idea is to implement this spec on Catapult v2, substituting "helm chart" with "test environment". As a minimum, Catapult v2 would create test environments:
   - env of kubectl, cf, helm
   - plus a cluster deployment
   - plus needed beforeTest configuration of that cluster and env (DNS, etc)
   - plus which specific charts are being tested, which targets run, etc)
This configuration would be declarative, in a folder of example values yaml, which would create a suite of test cases for CAP or others.

viccuad commented:
I have two concerns with the current draft:
- Namespace of the helm chart is not defined. We should define it either implicitly (`ns == name-of-chart`, eg: `kubecf`) or explicitly on a file. Which sounds similar to Helmfiles.
- Having all charts implement the spec means that development may become slower, if we need to bump spec version implementations across all of them.

prabalsharma commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,256 @@
+chart or caps-registry chart. But sometimes they develop automation that I could reuse.
+If they implement this spec, I will have access to their automation in simple,
+pull-way manner.
+
```
Maybe we should also add CAP / CAP CI user story, here.

viccuad replied:
The "downstream" developer user story is written with that intent. Would that be fitting?

mook-as commented:
How would the scripts (e.g. in `configure`) work? This defines no method of passing in options (e.g. to set the system domain, which presumably the chart itself can't know ahead of time, otherwise we'd just hard code it into the chart).

viccuad commented:
> How would the scripts (e.g. in `configure`) work? This defines no method of passing in options (e.g. to set the system domain, which presumably the chart itself can't know ahead of time, otherwise we'd just hard code it into the chart).

Currently in Catapult, and since cf-ci times, we use a configmap `cap-values` with information of the cluster external to Kube level.

Added specification for it in https://github.com/cloudfoundry-incubator/kubecf/pull/1495/commits/30fc89593119ab6ac7395af39bdca4d193d70c79.

mook-as commented:
> Added specification for it in [30fc895](https://github.com/cloudfoundry-incubator/kubecf/commit/30fc89593119ab6ac7395af39bdca4d193d70c79).

Does this mean the configuration is only available in the `configure` step? Also, how will it figure out _which_ config map to use (since we didn't name it in the spec)?  Should we have a spec-version in the config map too?

viccuad commented:
> Does this mean the configuration is only available in the `configure` step?

Correct, that's the idea, as there shouldn't be any need to read configurations on the other steps.

> Also, how will it figure out _which_ config map to use (since we didn't name it in the spec)?

Doh, I named it in the comment but not on the spec. Fixed in https://github.com/cloudfoundry-incubator/kubecf/pull/1495/commits/44c7fcdb6286bf139419f8a10ab2b5dd612a8e1d.

> Should we have a spec-version in the config map too?
Maybe we should have one, but since cluster creation is not part of the spec, and we haven't had the need to change the configmap keys in a long time, maybe it can wait for future spec bumps.

jandubois commented:
I've read this document several times, but it is still not clear to me how this would be used, and what benefits it will provide over what we already have.

I think it would be useful to know the requirements, the *what* you are trying to achieve; it is kind of hard to give feedback on the *how* without understand the *why* behind the decisions.

It sounds like the purpose is to simplify CAP QA, but the title is very generic "deployment and automation scripts", so maybe the scope is a lot narrower than I initially presumed. Is there more than a single consumer of the implementation of this spec?

mook-as commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,266 @@
+Deployment and automation scripts must expose the following targets, with the
+defined behaviour:
+
+- `chart`: Obtains the latest public release of the chart at hand, and puts it
```
We no longer need to call this `chart`; it can just be `download` at this point, since nothing else depends on this being a chart.

mook-as commented on doc/rfd/rfd/0009/README.md:
```diff
@@ -0,0 +1,266 @@
+  implemented, the intent is that test runners may decide to not call it and
+  drop themselves the chart under test in `$pwd/$our-chart/chart/`.
+
+- `configure`: Generates any config values that cannot be taken from the
```
We should also pass in what configuration is under test. (As in, the things under `example-values/`.)

viccuad commented:
> I think it would be useful to know the requirements, the _what_ you are trying to achieve; it is kind of hard to give feedback on the _how_ without understand the _why_ behind the decisions.
> 
> It sounds like the purpose is to simplify CAP QA, but the title is very generic "deployment and automation scripts", so maybe the scope is a lot narrower than I initially presumed. Is there more than a single consumer of the implementation of this spec?

You are not the first, so I think I failed to make my point in the "context" section of the spec.

The purpose is to simplify and reuse any CI that we do for any chart. Those that:
- Validate a new KubeCF release, against pre-existing charts (quarks, minibroker, stratos)
- Validate a new CAP release (formed from a set of charts) against pre-existing charts (previous CAP release)
- Validate a new Stratos release, against pre-existing charts (KubeCF, minibroker)
- Etc.

The idea is to:
- Get a declarative state of the system under test. This includes the values.yaml being used, and the cluster.
- Package CI scripts in a way that they can be consumed and modified downstream.
- Provide a versioned spec to iterate on the implementation.

viccuad commented:
We held a meeting to talk about the spec, and we extracted some points.

Concerns (in any order):
- Over engineered. Targets are too monolithic, one cannot inject operations in the middle by using the alphanumeric order anyways as one doesn't know which number to use, nor it may be needed.
- Set of values.yaml under test are not defined, and they may overlap when merging. The test runner has the responsibility of selecting the values under test, which lies outside of the spec, and renders the spec moot.
- May not be relevant for future implementations.
- Chart-specific implementation.
- Other consumers (E.g: Harbor) may not implement the spec. In that case, where does the script bundle live and how do we consume it?
- We don't have granularity on the testsuites to run. If we were to specify each testsuite, then the targets would be product specific.

Ideas:
- Spec object is shared between CIs, instead of sharing charts. The CIs would then use the spec object (from now on the "script bundle") to fetch the corect charts under test.
- Change the `chart` target to obtain the chart under test from either local path, url, or end-user repository.
- Simplify targets away from folders and alphanumeric execultables and into 4-5 scripts: clean-clobber, chart, install, upgrade.
- Don't have example-values be subsets of config.yamls to be merged, provide only sets of configs that can be fully tested.

Outcome:

From the ideas, it seemed that we need to define the usage and interface between spec objects and CI/users as a starting point.
A spike was created for it: https://github.com/cloudfoundry-incubator/kubecf/issues/1578.

A rewrite of the spec, simplifying it with the ideas and concerns, is needed.

viccuad commented:
Thulio shared the other day porter.sh; adopting it may very well satisfy this RFD.

jandubois commented:
@viccuad I've marked this PR as "stale". Please consider closing it for the time being (but keep the branch around for now), as I don't think any work around this will be scheduled in the near future.

viccuad commented:
Closing for now. From the planning, it would need a rewrite anyways.
